utils/remote-run

```diff
@@ -39,17 +44,48 @@ class CommandRunner(object):
         return subprocess.Popen(command, **kwargs)
 
     def send(self, local_to_remote_files):
-        # Prepare the remote directory structure.
-        # FIXME: This could be folded into the sftp connection below.
-        dirs_to_make = self._dirnames(local_to_remote_files.values())
-        self.run_remote(['/bin/mkdir', '-p'] + dirs_to_make)
-
-        # Send the local files.
-        sftp_commands = ("-put {0} {1}".format(quote(local_file), 
-                                               quote(remote_file))
-                         for local_file, remote_file
-                         in local_to_remote_files.items())
-        self.run_sftp(sftp_commands)
+        # Use tar to transfer the files we want
+        tmpdirname = tempfile.mkdtemp()
+        try:
+            archivename = os.path.join(tmpdirname, 'sent.tar.bz2')
+            with tarfile.open(name=archivename, mode='w:bz2') as archive:
+                for local_file, remote_file in local_to_remote_files.items():
+                    try:
+                        archive.add(local_file, arcname=remote_file)
+                    except OSError as e:
+                        if e.errno == errno.ENOENT:
+                            print("warning: {} not found".format(local_file),
+                                  file=sys.stderr)
+
+                            # In this case, we need to add the containing
+                            # directory to the tar file if we haven't already
+                            dirname = posixpath.dirname(remote_file)
+                            try:
+                                dirinfo = archive.getmember(dirname)
+                            except KeyError:
+                                dirinfo = tarfile.TarInfo(posixpath.dirname(remote_file))
+                                dirinfo.type = tarfile.DIRTYPE
+                                dirinfo.mode = 0o755
+                                archive.addfile(dirinfo)
+                        else:
+                            raise
+
+            invocation = self.remote_invocation(['/usr/bin/tar', '-x', '-j', '-P',
```

Comment:
I tend to prefer to write this as:

`target jxf - -P -C /` as `j`, `x`, `f` are the original Unix arguments and do not actually require the `-` as per Unix style argument passing.